crates/extensions/tedge_mqtt_bridge/src/lib.rs

```diff
@@ -0,0 +1,402 @@
+/// 6. The cloud MQTT end-point acknowledges the message, providing its cloud `pkid`.
+/// 7. The pair (cloud `pkid`, local message) is extracted from the cache and the local message is finaly acknowledged.
+///  
+/// ## Bridging cloud messages to the local broker
+///
+/// The very same two `half-bridge` instances ensure the reverse flow. Their roles are simply swapped:
+///
+/// - The `half_bridge(cloud_event_loop,local_client)` receives cloud messages and publishes these message locally.
+/// - The `half_bridge(local_event_loop,cloud_client)` handles the acknowledgements: waiting for messages be acknowledged locally, before sending acks for the original messages.
+///
+/// # Health topics
+/// The bridge will publish health information to `health_topic` (if supplied) on `target` to enable
+/// other components to establish bridge health. This is intended to be used the half with cloud
+/// event loop, so the status of this connection will be relayed to a relevant `te` topic like its
+/// mosquitto-based predecessor. The payload is either `1` (healthy) or `0` (unhealthy). When the
+/// connection is created, the last-will message is set to send the `0` payload when the connection
+/// is dropped.
+async fn half_bridge<F: for<'a> Fn(&'a str) -> Cow<'a, str>>(
+    mut recv_event_loop: EventLoop,
+    target: AsyncClient,
+    transform_topic: F,
+    mut companion_bridge_half: BidirectionalChannelHalf<Option<Publish>>,
+    health_topic: Option<String>,
+    name: &'static str,
+) {
+    let mut forward_pkid_to_received_msg = HashMap::new();
+    let mut bridge_health = BridgeHealth::new(name, health_topic, &target);
+
+    loop {
+        let res = recv_event_loop.poll().await;
+        bridge_health.update(&res, &mut companion_bridge_half).await;
+
+        let notification = match res {
+            Ok(notification) => notification,
+            Err(_) => continue,
+        };
+
+        match notification {
+            // Forward messages from event loop to target
+            Event::Incoming(Incoming::Publish(publish)) => {
+                target
+                    .publish(
+                        transform_topic(&publish.topic),
+                        publish.qos,
+                        publish.retain,
+                        publish.payload.clone(),
+                    )
+                    .await
+                    .unwrap();
+                let publish = (publish.qos > QoS::AtMostOnce).then_some(publish);
+                companion_bridge_half.send(publish).await.unwrap();
+            }
+
+            // Forward acks from event loop to target
+            Event::Incoming(
+                Incoming::PubAck(PubAck { pkid: ack_pkid })
+                | Incoming::PubRec(PubRec { pkid: ack_pkid }),
+            ) => {
+                if let Some(msg) = forward_pkid_to_received_msg.remove(&ack_pkid) {
+                    target.ack(&msg).await.unwrap();
```

Comment:
Shouldn't `target.ack(&msg).await.unwrap()` be done for `Incoming::PubComp` as well? Or is that done automatically if we don't have an explicit message handler here?